command/debug/debug.go

```diff
@@ -0,0 +1,702 @@
+		}
+
+		if _, err := io.Copy(t, f); err != nil {
+			return fmt.Errorf("failed to copy files for archive: %s", err)
+		}
+
+		f.Close()
+
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to walk output path for archive: %s", err)
+	}
+
+	// Remove directory that has been archived
+	err = os.RemoveAll(c.output)
+	if err != nil {
+		return fmt.Errorf("failed to remove archived directory: %s", err)
+	}
+
+	return nil
+}
+
+// defaultTargets specifies the list of all targets that
+// will be captured by default
+func (c *cmd) defaultTargets() []string {
+	return append(c.dynamicTargets(), c.staticTargets()...)
+}
+
+// dynamicTargets returns all the supported targets
+// that are retrieved at the interval specified
+func (c *cmd) dynamicTargets() []string {
+	return []string{"metrics", "logs", "pprof"}
+}
+
+// staticTargets returns all the supported targets
+// that are retrieved at the start of the command execution
+func (c *cmd) staticTargets() []string {
+	return []string{"host", "agent", "cluster"}
+}
+
+func (c *cmd) Synopsis() string {
+	return synopsis
+}
+
+func (c *cmd) Help() string {
+	return c.help
+}
+
+const synopsis = "Records a debugging archive for operators"
+const help = `
+Usage: consul debug [options]
+
+  Monitors a Consul agent for the specified period of time, recording
+  information about the agent, cluster, and environment to an archive
+  written to the specified path.
+
+  If ACLs are enabled, an 'operator:read' token must be supplied in order
+  to perform this operation.
+
+  To create a debug archive in the current directory for the default
+  duration and interval, capturing all information available:
+
+      $ consul debug
+
+  The command stores captured data at the configured output path
+  through the duration, and will archive the data at the same
+  path if interrupted.
+
+  Flags can be used to customize the duration and interval of the
+  operation. Duration is the total time to capture data for from the target
+  agent and interval controls how often dynamic data such as metrics
+  are scraped.
+
+      $ consul debug -interval=20s -duration=1m
+
+  The capture flag can be specified multiple times to limit information
+  retrieved.
+
+      $ consul debug -capture metrics -capture agent
+
+  By default, the archive containing the debugging information is
+  saved to the current directory as a .tar.gz file. The
+  output path can be specified, as well as an option to disable
+  archiving, leaving the directory intact.
+
+      $ consul debug -output=/foo/bar/my-debugging -archive=false
+
+  Note: Information collected by this command has the potential
+  to be highly sensitive. Sensitive material such as ACL tokens and
+  other commonly secret material are redacted automatically, but we
+  strongly recommend review of the data within the archive prior to
+  transmitting it.
+
+  For a full list of options and examples, please see the Consul
+  documentation.
+`
```

Comment:
This looks great. Is it important to note that for pprof stuff it needs to be enabled or do we just show a useful error if it's not?

Reply:
It'll just show a useful `WARN` message if not, and skip pprof.